### PR TITLE
Wrap data in set_data_io with a DataChunkIterator to support overriding hdf5 dataset backend configurations

### LINKED ISSUE #1170: [Feature]: Add support for overriding backend configuration in HDF5 datasets
### What would you like to see added to HDMF?

I am working on a new helper feature for neuroconv, in which users can repack an NWB file with new backend configurations (https://github.com/catalystneuro/neuroconv/pull/1003). However when I try to export the NWB file with the new backend configurations, I get a user warning and the new backend configuration is ignored.
```bash
/opt/anaconda3/envs/neuroconv_tdtfp_env/lib/python3.12/site-packages/hdmf/utils.py:668: UserWarning: chunks in H5DataIO will be ignored with H5DataIO.data being an HDF5 dataset
````

### What solution would you like?

I was able to solve this problem by simply converting the HDF5 dataset to a numpy array like so:
```diff
  # hdmf.container.Container
  def set_data_io(self, dataset_name: str, data_io_class: Type[DataIO], data_io_kwargs: dict = None, **kwargs):
          """
          Apply DataIO object to a dataset field of the Container.
  
          Parameters
          ----------
          dataset_name: str
              Name of dataset to wrap in DataIO
          data_io_class: Type[DataIO]
              Class to use for DataIO, e.g. H5DataIO or ZarrDataIO
          data_io_kwargs: dict
              keyword arguments passed to the constructor of the DataIO class.
          **kwargs:
              DEPRECATED. Use data_io_kwargs instead.
              kwargs are passed to the constructor of the DataIO class.
          """
          if kwargs or (data_io_kwargs is None):
              warn(
                  "Use of **kwargs in Container.set_data_io() is deprecated. Please pass the DataIO kwargs as a "
                  "dictionary to the `data_io_kwargs` parameter instead.",
                  DeprecationWarning,
                  stacklevel=2
              )
              data_io_kwargs = kwargs
          data = self.fields.get(dataset_name)
+         data = np.array(data)
          if data is None:
              raise ValueError(f"{dataset_name} is None and cannot be wrapped in a DataIO class")
          self.fields[dataset_name] = data_io_class(data=data, **data_io_kwargs)
```

I would appreciate some kind of alternative set_data_io() function that supports overwriting HDF5 data sets in this manner (or something similar).

### Do you have any interest in helping implement the feature?

Yes.

### PR DESCRIPTION
**Pull request by pauladkisson**
## Motivation

Fixes #1170 

## Checklist

- [x] Did you update `CHANGELOG.md` with your changes?
- [x] Does the PR clearly describe the problem and the solution?
- [x] Have you reviewed our [Contributing Guide](https://github.com/hdmf-dev/hdmf/blob/dev/docs/CONTRIBUTING.rst)?
- [x] Does the PR use "Fix #XXX" notation to tell GitHub to close the relevant issue numbered XXX when the PR is merged?

### CODE REVIEW

**Comment by pauladkisson** [comment]
@oruebel, are there any problems with wrapping _all_ data in a DataChunkIterator (numpy arrays, lists, tuples)? Should this fix be restricted to hdf5 datasets?

**Comment by oruebel** [comment]
> are there any problems with wrapping _all_ data in a DataChunkIterator (numpy arrays, lists, tuples)? Should this fix be restricted to hdf5 datasets?

Since the user has explicit control, I think this is fine. Also,, I believe it should work with numpy, lists, and tuples as well.

File: src/hdmf/container.py [hunk 1]
```diff
@@ -830,7 +830,13 @@ def __smart_str_dict(d, num_indent):
...
-    def set_data_io(self, dataset_name: str, data_io_class: Type[DataIO], data_io_kwargs: dict = None, **kwargs):
+    def set_data_io(
+        self,
+        dataset_name: str,
+        data_io_class: Type[DataIO], data_io_kwargs: dict = None,
+        data_chunk_iterator_class: Type[AbstractDataChunkIterator] = DataChunkIterator,
+        data_chunk_iterator_kwargs: dict = None, **kwargs
+    ):
         """
         Apply DataIO object to a dataset field of the Container.
```
Comment (oruebel) [inline src/hdmf/container.py:839]: Can you add a comment here explaining why one would want to wrap with `DataChunkIterator` in addition to the `H5DataIO`

File: src/hdmf/container.py [hunk 2]
```diff
@@ -830,7 +830,13 @@ def __smart_str_dict(d, num_indent):
         out += '\n' + indent + right_br
         return out
 
-    def set_data_io(self, dataset_name: str, data_io_class: Type[DataIO], data_io_kwargs: dict = None, **kwargs):
+    def set_data_io(
+        self,
+        dataset_name: str,
+        data_io_class: Type[DataIO], data_io_kwargs: dict = None,
+        data_chunk_iterator_class: Type[AbstractDataChunkIterator] = DataChunkIterator,
```
Comment (oruebel) [inline src/hdmf/container.py]:
```suggestion
        data_io_class: Type[DataIO], 
        data_io_kwargs: dict = None,
        data_chunk_iterator_class: Type[AbstractDataChunkIterator] = DataChunkIterator,
```

File: src/hdmf/container.py [hunk 3]
```diff
@@ -855,8 +865,10 @@ def set_data_io(self, dataset_name: str, data_io_class: Type[DataIO], data_io_kw
             )
             data_io_kwargs = kwargs
         data = self.fields.get(dataset_name)
+        data_chunk_iterator_kwargs = data_chunk_iterator_kwargs or dict()
         if data is None:
             raise ValueError(f"{dataset_name} is None and cannot be wrapped in a DataIO class")
+        data = data_chunk_iterator_class(data=data, **data_chunk_iterator_kwargs)
```
Comment (oruebel) [inline src/hdmf/container.py]: I just realized using `DataChunkIterator` as a default argument is not a good idea. In most cases a user will just want to set `DataIO` and not apply a `DataChunkIterator`, but the current behavior here is to always set a `DataChunkIterator`. I.e., I would suggest changing the default for the `data_chunk_iterator_class` to `None` and then here do 
```suggestion
        if data_chunk_iterator_class is not None:
            data = data_chunk_iterator_class(data=data, **data_chunk_iterator_kwargs)
```

File: src/hdmf/container.py [hunk 4]
```diff
@@ -910,8 +928,14 @@ def set_data_io(self, data_io_class: Type[DataIO], data_io_kwargs: dict) -> None
...
         data_io_kwargs: dict
             The keyword arguments to pass to the DataIO.
+        data_chunk_iterator_class: Type[AbstractDataChunkIterator]
+            The DataChunkIterator to use for the DataIO, by default DataChunkIterator.
+        data_chunk_iterator_kwargs: dict
+            The keyword arguments to pass to the DataChunkIterator.
         """
-        self.__data = data_io_class(data=self.__data, **data_io_kwargs)
+        data_chunk_iterator_kwargs = data_chunk_iterator_kwargs or dict()
+        data = data_chunk_iterator_class(data=self.__data, **data_chunk_iterator_kwargs)
```
Comment (oruebel) [inline src/hdmf/container.py]: Same comment as above. Wrapping with data_chunk_iterator_class should be optional.

**Comment by pauladkisson** [comment]
@oruebel, can you enable tests for this PR?

**Comment by pauladkisson** [comment]
@oruebel @rly CI tests pls?

**Comment by pauladkisson** [comment]
Looks like this is passing all the tests, can we merge?

**Comment by pauladkisson** [comment]
@rly @oruebel

**Comment by rly** [comment]
This looks good to me.

**Review by rly — DISMISSED** [review]

**Review by rly — APPROVED** [review]

**Comment by rly** [comment]
Thank you @pauladkisson !

**Pull request merged**